File: openFaces/source/org/openfaces/renderkit/validation/FloatingIconMessageRenderer.js
/*
 * OpenFaces - JSF Component Library 2.0
 * Copyright (C) 2007-2009, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
O$._FloatingIconMessageRenderer = function(clientId, forClientId, imageUrl, topOffset, leftOffset, css, hideImg, showSummary, showDetail, defaultPresentation) {
  this.clientId = clientId;
  var messageElement = O$(this.clientId);
  if (messageElement) {
    var parentNodeElement = messageElement.parentNode;
    parentNodeElement.removeChild(messageElement);
  }
  this.forId = forClientId;
  this.imageUrl = imageUrl;
  this.topOffset = topOffset;
  this.leftOffset = leftOffset;
  this.errorCss = css;
  this.showImg = !hideImg;
  this.showSummary = showSummary;
  this.showDetail = showDetail;
  this.defaultPresentation = defaultPresentation;
}

O$._FloatingIconMessageRenderer.prototype.renderMessage = function(message) {
  if (this.defaultPresentation && O$._isComponentHasPresentation(this.forId)) {
    return;
  }
  var forElement = O$.byIdOrName(this.forId);

  if (message) {
    var messageText = "";
    if (this.showSummary) messageText = message.summary;
    if (this.showDetail) messageText = message.detail;
    if (this.showDetail && this.showSummary)messageText = message.summary + "; " + message.detail;
    if (!this.showDetail && !this.showSummary) messageText = message.summary;

    if (this.showImg /*|| (!this.showImg && !this.errorCss)*/) {
      var messageElement = O$(this.clientId);
      if (!messageElement) {
        messageElement = document.createElement("img");
        messageElement.id = this.clientId;
        messageElement.src = this.imageUrl;
        messageElement.style.position = 'absolute';
        messageElement.style.zIndex = 700;
        document.body.appendChild(messageElement);
        O$.getClientMessageRenderersWithVisibleBubble()[this.clientId + '_' + this.forId] = this;
      }
      this.position();
      messageElement.alt = messageText;
      messageElement.title = messageText;

    }
    if (!forElement._of_titleChanged) {
      forElement._of_title = forElement.title;
      forElement._of_titleChanged = true;
    }
    //    forElement.title = message.detail;
    forElement.title = messageText;

    if (this.errorCss && !forElement._of_cssChanged) {
      forElement._of_className = forElement.className;
      forElement.className = forElement.className + " " + this.errorCss;
      forElement._of_cssChanged = true;
      O$.repaintAreaForOpera(forElement, true);
    }

  } else {
    var messageElement = O$(this.clientId);
    if (messageElement) {
      var parentNode = messageElement.parentNode;
      parentNode.removeChild(messageElement);
    }
    if (forElement._of_titleChanged) {
      forElement.title = forElement._of_title;
      forElement._of_title = undefined;
      forElement._of_titleChanged = undefined;
    }
    if (forElement._of_cssChanged) {
      forElement.className = forElement._of_className;
      forElement._of_className = undefined;
      forElement._of_cssChanged = undefined;
      O$.repaintAreaForOpera(forElement, true);
    }
  }
}

O$._FloatingIconMessageRenderer.prototype.update = function() {
  var forElement = O$.byIdOrName(this.forId);
  //changed logic of message rendering
  // now we need to check for component until we'll find it, required for Ajax Support
  //  if (forElement == null) {
  //    setTimeout(function() {
  //      this.update();
  //      alert("This update");
  //    }, 10);
  //    return;
  //  }
  var message;
  if (forElement) {
    message = O$.getMessages(forElement)[0];
    this.renderMessage(message);
  }
}

O$._FloatingIconMessageRenderer.prototype.position = function() {
  var messageElement = O$(this.clientId);
  if (messageElement) {
    var forElement = O$.byIdOrName(this.forId);

    if (!O$.isVisibleRecursive(forElement)) {
      messageElement.style.display = "none";
      return;
    }
    messageElement.style.display = "";

    var elementPos = O$.getElementPos(forElement);
    var left = elementPos.left + this.leftOffset;
    var top = elementPos.top + this.topOffset;

    messageElement.style.left = left + "px";
    messageElement.style.top = top + "px";
  }
}

File: openFaces/source/org/openfaces/renderkit/validation/FloatingIconMessageRenderer.java
/*
 * OpenFaces - JSF Component Library 2.0
 * Copyright (C) 2007-2010, TeamDev Ltd.
 * licensing@openfaces.org
 * Unless agreed in writing the contents of this file are subject to
 * the GNU Lesser General Public License Version 2.1 (the "LGPL" License).
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * Please visit http://openfaces.org/licensing/ for more details.
 */
package org.openfaces.renderkit.validation;

import org.openfaces.component.validation.ClientValidationMode;
import org.openfaces.component.validation.FloatingIconMessage;
import org.openfaces.component.validation.ValidationProcessor;
import org.openfaces.org.json.JSONObject;
import org.openfaces.util.Components;
import org.openfaces.util.NewInstanceScript;
import org.openfaces.util.Rendering;
import org.openfaces.util.Resources;
import org.openfaces.util.Script;
import org.openfaces.util.ScriptBuilder;
import org.openfaces.util.Styles;
import org.openfaces.validator.ClientValidatorUtil;

import javax.faces.application.FacesMessage;
import javax.faces.component.UIComponent;
import javax.faces.component.UIForm;
import javax.faces.context.FacesContext;
import java.io.IOException;
import java.util.Iterator;

/**
 * @author Vladimir Korenev
 */
public class FloatingIconMessageRenderer extends BaseMessageRenderer {
    private static final String DEFAULT_CLASS = "o_floatingIconMessage";

    public void encodeEnd(FacesContext context, UIComponent component) throws IOException {
        super.encodeEnd(context, component);

        if (!component.isRendered())
            return;

        FloatingIconMessage fim = (FloatingIconMessage) component;

        UIComponent forComponent = getForComponent(fim);
        boolean pageDefinedMessage = !fim.isRuntimeDefined();
        ValidationProcessor validationProcessor = ValidationProcessor.getInstance(context);
        if (validationProcessor == null)
            return;


        ClientValidationMode cv = validationProcessor.getClientValidationRule(fim, forComponent);
        UIForm form = Components.getEnclosingForm(fim);
        boolean clientValidation = !cv.equals(ClientValidationMode.OFF);
        boolean useDCVP = validationProcessor.isUseDefaultClientValidationPresentationForForm(form);
        boolean useDSVP = validationProcessor.isUseDefaultServerValidationPresentationForForm(form);

        String forComponentClientId = getForComponentClientId(context, fim);
        if (forComponentClientId == null) {
            Rendering.logWarning(context, "Cannot render floatingIconMessage bacause can't calculate " +
                    "target component client ID. It may be caused by 'for' attribute absence");
            return;
        }
        Script clientScript = getClientScript(context, fim, forComponentClientId,
                clientValidation, pageDefinedMessage, useDCVP, useDSVP);
        Rendering.renderInitScript(context, clientScript, getJavascriptLibraryUrls(context));
        if (clientScript.toString().length() > 0) {
            if (!isDefaultPresentation(fim))
                ValidatorUtil.renderPresentationExistsForComponent(forComponentClientId, context);
            Styles.renderStyleClasses(context, fim, true, false);
        }
    }

    private Script getClientScript(FacesContext context,
                                   FloatingIconMessage fim,
                                   String forComponentClientId,
                                   boolean clientValidation,
                                   boolean pageDefinedMessage,
                                   boolean useDefaultClientPresentation,
                                   boolean useDefaultServerPresentation) {
        if (forComponentClientId == null) {
            throw new NullPointerException("forComponentClientId");
        }
        ScriptBuilder resultScript = new ScriptBuilder();
        Iterator messages = context.getMessages(forComponentClientId);
        FacesMessage message = null;

        if (messages.hasNext()) {
            message = (FacesMessage) messages.next();
        }

        String css = Styles.getCSSClass(context, fim,
                !fim.isNoStyle() ? fim.getStyle() : null, DEFAULT_CLASS,
                !fim.isNoStyle() ? fim.getStyleClass() : null);
        JSONObject events = Rendering.getEventsParam(fim, "onclick", "ondblclick", "onmousedown", "onmouseover", "onmousemove", "onmouseout", "onmouseup");

        ScriptBuilder clientValidationScript = new ScriptBuilder().functionCall("O$.addClientMessageRenderer",
                new NewInstanceScript("O$._FloatingIconMessageRenderer",
                        fim,
                        forComponentClientId,
                        getImageUrl(fim),
                        getOffsetTop(fim),
                        getOffsetLeft(fim),
                        css,
                        events,
                        fim.isNoImage(),
                        fim.isShowSummary(),
                        fim.isShowDetail(),
                        isDefaultPresentation(fim))).semicolon();

        ScriptBuilder serverValidationScript = new ScriptBuilder();
        if (message != null) {
            serverValidationScript.onLoadScript(new ScriptBuilder().append(
                    ClientValidatorUtil.getScriptAddMessageById(message, forComponentClientId)).

                    newInstance("O$._FloatingIconMessageRenderer",
                            fim,
                            forComponentClientId,
                            getImageUrl(fim),
                            getOffsetTop(fim),
                            getOffsetLeft(fim),
                            css,
                            events,
                            fim.isNoImage(),
                            fim.isShowSummary(),
                            fim.isShowDetail(),
                            isDefaultPresentation(fim)).append(".update();\n"));
        }

        if (pageDefinedMessage) {
            if (clientValidation) {
                resultScript = new ScriptBuilder(clientValidationScript);
            }
            if (message != null) {
                resultScript.append(serverValidationScript);
            }
        } else {
            if (clientValidation && useDefaultClientPresentation) {
                resultScript = new ScriptBuilder(clientValidationScript);
            }
            if (useDefaultServerPresentation) {
                if (message != null) {
                    resultScript.append(serverValidationScript);
                } else {
                    if (clientValidation) {
                        resultScript = new ScriptBuilder(clientValidationScript);
                    }
                }
            }
        }
        return resultScript;
    }

    private String[] getJavascriptLibraryUrls(FacesContext context) {
        return new String[]{
                Resources.getUtilJsURL(context),
                Resources.getInternalURL(context, this.getClass(), "floatingIconMessage.js"),
                ValidatorUtil.getValidatorUtilJsUrl(context)
        };
    }

    protected String getImageUrl(UIComponent component) {
        if (component instanceof FloatingIconMessage) {
            return ((FloatingIconMessage) component).getImageUrl();
        } else {
            return (String) component.getAttributes().get("imageUrl");
        }
    }


    protected int getOffsetTop(UIComponent component) {
        if (component instanceof FloatingIconMessage) {
            return ((FloatingIconMessage) component).getOffsetTop();
        } else {
            Number n = (Number) component.getAttributes().get("offsetTop");
            return n.intValue();
        }
    }

    protected int getOffsetLeft(UIComponent component) {
        if (component instanceof FloatingIconMessage) {
            return ((FloatingIconMessage) component).getOffsetLeft();
        } else {
            Number n = (Number) component.getAttributes().get("offsetLeft");
            return n.intValue();
        }
    }
}
